core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -189,6 +233,15 @@ static Expression joinSelector(JoinRelType joinType, PhysType physType,
         expressions.add(expression);
       }
     }
+
+    if (generateCompactCode) {
```

Comment:
if you could add a comment showing here the shape of the generated code it could help people understand easier what is going on.

Reply:
@thomasrebele 

Here is the code generated from a unit test.
```
{
  return org.apache.calcite.linq4j.EnumerableDefaults.nestedLoopJoin(org.apache.calcite.schema.Schemas.queryable(root, root.getRootSchema().getSubSchema("ROOT"), org.apache.calcite.interpreter.Row.class, "T1").asEnumerable(), org.apache.calcite.schema.Schemas.queryable(root, root.getRootSchema().getSubSchema("ROOT"), org.apache.calcite.interpreter.Row.class, "T0").asEnumerable(), new org.apache.calcite.linq4j.function.Predicate2() {
      public boolean apply(org.apache.calcite.interpreter.Row left, org.apache.calcite.interpreter.Row right) {
        return true;
      }
      public boolean apply(Object left, Object right) {
        return apply(
          (org.apache.calcite.interpreter.Row) left,
          (org.apache.calcite.interpreter.Row) right);
      }
    }
    , new org.apache.calcite.linq4j.function.Function2() {
      public String[] apply(org.apache.calcite.interpreter.Row left, org.apache.calcite.interpreter.Row right) {
        String[] outputArray = new String[201];
        return outputArray;
      }
      public String[] apply(Object left, Object right) {
        return apply(
          (org.apache.calcite.interpreter.Row) left,
          (org.apache.calcite.interpreter.Row) right);
      }
    }
    , org.apache.calcite.linq4j.JoinType.INNER);
}
```

This is just returning an initialized String array of the correct size. The data isn't copied.

The issue could be from here:
https://github.com/Bit-Quill/calcite/blob/calcite-3094-method-too-large/core/src/main/java/org/apache/calcite/adapter/enumerable/JavaRowFormat.java#L347

What should `JavaRowFormat.copy()` return?

Reply:
Indeed, I had not implemented `JavaRowFormat.copy()`, as I didn't know how to do that. Seems that the input would be `org.apache.calcite.interpreter.Row` according to the generated code. I'm astonished that the return type is not of the same type.
The `JavaRowFormat.copy()` would need to return an `Expression` that represents the code to copy the data from the left and right `org.apache.calcite.interpreter.Row`. Not sure whether an `ARRAY_COPY` could be used here. Alternatively, the generated code could be a for loop, here some pseudo-code:
```
int outIdx = 0;
for(int leftIdx = 0; leftIdx < left length; leftIdx++) {
  outputArray[outIdx++] = left item at position leftIdx;
}
for(int rightIdx = 0; rightIdx < right length; rightIdx++) {
  outputArray[outIdx++] = right item at position rightIdx;
}
```

Reply:
I've changed copy to generate the following code:

`new org.apache.calcite.linq4j.function.Function2() {
  public String[] apply(org.apache.calcite.interpreter.Row left, org.apache.calcite.interpreter.Row right) {
    String[] outputArray = new String[201];
    System.arraycopy(left.copyValues(), 0, outputArray, 0, 101);
    System.arraycopy(right.copyValues(), 0, outputArray, 101, 100);
    return outputArray;
  }
  public String[] apply(Object left, Object right) {
    return apply(
      (org.apache.calcite.interpreter.Row) left,
      (org.apache.calcite.interpreter.Row) right);
  }
}
`

I'll add that to the comments.

Reply:
@mihaibudiu I think your comments are now addressed. Thanks!